src/components/intent.ts

```diff
@@ -827,23 +850,87 @@ export class Intent {
+                type: "m.id.user",
+                user: userId,
+            }
+        });
+        return {
+            accessToken: res.access_token,
+            deviceId: res.device_id,
+        };
+    }
+
+    public async ensureRegistered() {
+        log.debug("Checking if user is registered");
+        if (this.opts.registered && !this.encryption) {
+            log.debug("ensureRegistered: Registered, and not encrypted");
             return "registered=true";
         }
-        const userId = this.client.credentials.userId;
-        const localpart = new MatrixUser(userId).localpart;
-        try {
-            const res = await this.botClient.register(localpart);
-            this.opts.registered = true;
-            return res;
-        }
-        catch (err) {
-            if (err.errcode === "M_USER_IN_USE") {
+        let registerRes;
+        const userId: string = this.client.credentials.userId;
+        if (!this.opts.registered) {
+            const localpart = new MatrixUser(userId).localpart;
+            try {
+                registerRes = await this.botClient.register(localpart);
+                this.opts.registered = true;
+            }
+            catch (err) {
+                if (err.errcode !== "M_USER_IN_USE") {
+                    throw err;
+                }
                 this.opts.registered = true;
-                return null;
             }
-            throw err;
         }
+
+        if (!this.encryption) {
+            log.debug("ensureRegistered: Registered, and not encrypted");
+            // We don't care about encryption, or the encryption is ready.
+            return registerRes;
+        }
+
+        if (this.readyPromise) {
+            log.debug("ensureRegistered: ready promise ongoing");
+            try {
+                // Should fall through and find the session.
+                await this.readyPromise;
+            }
+            catch (ex) {
+                log.debug("ensureRegistered: failed to ready", ex);
+                // Failed to ready up - fall through and try again.
+            }
+        }
+
+        // Encryption enabled, check if we have a session.
+        let session = await this.encryption.sessionPromise;
+        if (session) {
+            log.debug("ensureRegistered: Existing enc session, reusing");
+            // We have existing credentials, set them on the client and run away.
+            this.client._http.opts.accessToken = session.accessToken;
+        }
+        else {
+            this.readyPromise = (async () => {
+                log.debug("ensureRegistered: Attempting encrypted login");
+                // Login as the user
+                const result = await this.loginForEncryptedClient();
+                session = {
+                    userId,
+                    ...result,
+                };
+                if (this.encryption) {
+                    this.encryption.sessionPromise = Promise.resolve(session);
```

Comment:
I don't understand what's happening here.
The `encryption` object is coming from opts, so I guess the parent of this object will later check this promise and get the value???

Why is this a promise and not the value?

Reply:
Yes, this is a tricky one. In order to not break certain APIs, we have to ensure that the *creation* of an intent is not asynchronous. This means we can't do any store lookups, any API calls or anything until the intent is used. The problem with this is it means the first opportunity to do anything is in ensureRegistered, which is always called before we perform an intent action.

So things like the session which we need to pass to the client have to be passed as a promise to unwrap once we call ensureRegistered. Ideally we wouldn't be doing this at this stage, but it's that or breaking API compatibility, which felt like it would lead to a lot of work on the other bridges. On the whole passing a promise isn't the ultimate evil.